### PR TITLE
Some RocksDB API additions

### PR DESCRIPTION
**Pull request by delamonpansie**

### CODE REVIEW

File: rocks_common.ml [hunk 1]
```diff
@@ -84,8 +84,10 @@ let make_destroy t destructor =
       destructor
       (t @-> returning void) in
   fun t ->
-  inner t;
-  t.valid <- false
+  if t.valid then begin
```
Comment (domsj) [inline rocks_common.ml]: I'm still not sure if we want to be silent about the caller destroying a resource twice ... it indicates a bug that should be handled.
Do you still need this change?

File: rocks.ml [hunk 2]
```diff
@@ -215,13 +224,232 @@ module rec Iterator : Rocks_intf.ITERATOR with type db := RocksDb.t = struct
...
+    let inner opts =
+      let res_size = allocate Views.int_to_size_t 0 in
+      let res = with_err_pointer
+          (get_raw t opts (bigarray_start array1 key +@ pos) len res_size)
+      in
+      if (to_voidp res) = null
+      then None
+      else begin
+        let res' = bigarray_of_ptr array1 (!@res_size) Bigarray.char res in
+        Gc.finalise (fun res -> free (to_voidp res)) res;
```
Comment (domsj) [inline rocks.ml]: Does this `free` actually need to be in a finalizer?
(I know, in other places it's done like that too ... but I'm wondering if I made a mistake there)

**Review by domsj — COMMENTED** [review]
Thanks for the additions, a few comments below

File: rocks_common.ml (hunk 1 above)
Comment (delamonpansie) [reply rocks_common.ml]: I'm not sure should this be considered as a bug. Anyway, throwing an exception from finalyzer is a bad idea.
I have no code that calls finalyzer twice.

File: rocks.ml (hunk 2 above)
Comment (delamonpansie) [reply rocks.ml]: Honestly, I don't know. Since C API is a bit different I was forced to copy wrappers from your code.

Comment (domsj) [reply rocks.ml]: hmm, I guess for the `bigarray_of_ptr` it is required to wait with `free`ing until `res'` is no longer referenced, see http://ocamllabs.io/ocaml-ctypes/Ctypes.html#VALbigarray_of_ptr
I'm not sure `res'` will also keep `res` alive ... need to dig into ctypes a bit more again.

For the `string_from_ptr` variant I think we can free the original c-string immediately.

File: rocks.ml [hunk 3]
```diff
@@ -215,13 +224,232 @@ module rec Iterator : Rocks_intf.ITERATOR with type db := RocksDb.t = struct
...
+      "rocksdb_transaction_begin"
+      (db @-> WriteOptions.t @-> TransactionOptions.t @-> ptr void @->
+         returning t)
+
+  let destroy = make_destroy t destructor
+
+  let txnbegin ?wopts ?txnopts db =
+    let inner wopts txnopts =
+      let t = txnbegin_raw db wopts txnopts null in
+      Gc.finalise destroy t;
```
Comment (domsj) [inline rocks.ml]: This finalizer in combination with the `with_t` below is probably the source of your double freeing problems ...

**Pull request closed without merging**